graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java

```diff
@@ -0,0 +1,358 @@
+            }
+            return Collections.unmodifiableMap((Map<String, Object>) maybeMap);
+        } else {
+            throw new IllegalArgumentException("Unknown parameter type variables");
+        }
+    }
+
+    private <T> T parseJsonString(String content, TypeReference<T> typeReference)
+            throws JsonProcessingException {
+        return mapper.readValue(content, typeReference);
+    }
+
+    private void writePong(WebSocketWriter out) {
+        out.tryWrite("{\"type\":\"pong\"}");
+    }
+
+    private void writeConnectionAck(WebSocketWriter out) {
+        out.tryWrite("{\"type\":\"connection_ack\"}");
+    }
+
+    private void writeNext(WebSocketWriter out, String operationId, ExecutionResult executionResult) {
+        final HashMap<String, Object> response = new HashMap<>();
+        response.put("id", operationId);
+        response.put("type", "next");
+        response.put("payload", executionResult.toSpecification());
+        try {
+            final String event = serializeToJson(response);
+            logger.trace("NEXT: {}", event);
+            out.tryWrite(event);
+        } catch (JsonProcessingException e) {
+            logger.error("Error serializing next event", e);
+            writeError(out, operationId, e);
+        }
+    }
+
+    private void writeError(WebSocketWriter out, String operationId, List<GraphQLError> errors) {
+        final HashMap<String, Object> errorResponse = new HashMap<>();
+        errorResponse.put("type", "error");
+        errorResponse.put("id", operationId);
+        final List<Map<String, Object>> errorSpecifications =
+                errors.stream().map(GraphQLError::toSpecification).collect(Collectors.toList());
+        errorResponse.put("payload", errorSpecifications);
+        try {
+            final String event = serializeToJson(errorResponse);
+            logger.trace("ERROR: {}", event);
+            out.tryWrite(event);
+        } catch (JsonProcessingException e) {
+            logger.error("Error serializing error event", e);
+            writeError(out, operationId, e);
+        }
+    }
+
+    private void writeError(WebSocketWriter out, String operationId, Throwable t) {
+        final HashMap<String, Object> errorResponse = new HashMap<>();
+        errorResponse.put("type", "error");
+        errorResponse.put("id", operationId);
+        errorResponse.put("payload", ImmutableList.of(
+                new GraphQLError() {
+                    @Override
+                    public String getMessage() {
+                        return t.getMessage();
+                    }
+
+                    @Override
+                    public List<SourceLocation> getLocations() {
+                        return emptyList();
+                    }
+
+                    @Override
+                    public ErrorClassification getErrorType() {
+                        return ErrorClassification.errorClassification("Unknown");
```

Comment:
Question: Shouldn't we use the exception type instead? 🤔

Reply:
There is no obvious mapping between a generic `Throwable` and a GraphQL Error classification.

I updated to the built-in generic `ErrorType.DataFetchingException`